catalyst/utils/distributed.py

```diff
@@ -0,0 +1,237 @@
+        argspec.args[-len(argspec.defaults):], argspec.defaults
+    )
+    if exclude is not None:
+        default_params = filter(lambda x: x[0] not in exclude, default_params)
+    default_params = dict(default_params)
+    return default_params
+
+
+def get_slurm_params():
+    cmd = "scontrol show hostnames '%s'" % os.environ["SLURM_JOB_NODELIST"]
+    nodes = subprocess.getoutput(cmd).split()
+    num_nodes = int(os.environ["SLURM_JOB_NUM_NODES"])
+    current_node = os.environ["SLURMD_NODENAME"]
+    master_node = socket.gethostbyname(nodes[0])
+    cur_node_idx = nodes.index(current_node)
+    return cur_node_idx, num_nodes, master_node
+
+
+def is_slurm():
+    return "SLURM_JOB_NUM_NODES" in os.environ and "SLURM_NODEID" in os.environ
+
+
+def get_distributed_params():
+    master_addr = "127.0.0.1"
+    cur_node, num_nodes = 0, 1
+    if is_slurm():
+        cur_node, num_nodes, master_addr = get_slurm_params()
+
+    os.environ["MASTER_ADDR"] = os.getenv("MASTER_ADDR", master_addr)
+    os.environ["MASTER_PORT"] = os.getenv("MASTER_PORT", "424242")
+
+    workers_per_node = torch.cuda.device_count()
+    start_rank = cur_node * workers_per_node
+    world_size = num_nodes * workers_per_node
+
+    local_rank = os.getenv("LOCAL_RANK", None)
+    rank = os.getenv("RANK", None)
+    local_rank, rank = [v and int(v) for v in [local_rank, rank]]
+    world_size = int(os.getenv("WORLD_SIZE", world_size))
+
+    return OrderedDict(
+        local_rank=local_rank,
+        start_rank=start_rank,
+        rank=rank,
+        world_size=world_size,
+        master_addr=os.environ["MASTER_ADDR"],
+        master_port=os.environ["MASTER_PORT"],
+    )
+
+
+def get_distributed_env(local_rank, rank, world_size):
+    env = os.environ.copy()
+    env["LOCAL_RANK"] = str(local_rank)
+    env["RANK"] = str(rank)
+    env["WORLD_SIZE"] = str(world_size)
+    return env
+
+
+def distributed_run(data_parallel, worker_fn, *args, **kwargs):
+    distributed_params = get_distributed_params()
+    local_rank = distributed_params["local_rank"]
+    world_size = distributed_params["world_size"]
+
+    if data_parallel or world_size <= 1:
+        worker_fn(*args, **kwargs)
+    elif local_rank is not None:
+        torch.cuda.set_device(int(local_rank))
+
+        torch.distributed.init_process_group(
+            backend="nccl", init_method="env://"
+        )
+        worker_fn(*args, **kwargs)
+    else:
+        workers = []
+        try:
+            for local_rank in range(torch.cuda.device_count()):
+                rank = distributed_params["start_rank"] + local_rank
+                env = get_distributed_env(local_rank, rank, world_size)
+                cmd = [sys.executable] + sys.argv.copy()
+                workers.append(subprocess.Popen(cmd, env=env))
+            for worker in workers:
+                worker.wait()
+        finally:
+            for worker in workers:
+                worker.kill()
+
+
+def process_components(
```

Comment:
don't you mind to place this function back to `catalyst/utils/torch.py` ?

Reply:
`process_components` actually contains a wrapper for distributed models. I think it's better to place `process_components` and `distributed_run` in the same file. This will reduce potential errors, because when editing one of these functions, it is necessary to check the other.